src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs

```diff
@@ -159,28 +159,27 @@ private Task BindArgumentsAsync()
                 await CacheEntry.PropertyBinder(_pageContext, _instance);
             }
 
-            if (_handler == null)
+            if (_handler == null || _handler.Parameters.Count == 0)
             {
                 return;
             }
 
-            var valueProvider = await CompositeValueProvider.CreateAsync(_pageContext, _pageContext.ValueProviderFactories);
+            Debug.Assert(
+                _actionDescriptor.HandlerMethods.Count == CacheEntry.HandlerBinders.Length,
+                "Handler has parameters, but binders are not set up correctly.");
 
-            for (var i = 0; i < _handler.Parameters.Count; i++)
+            PageHandlerBinderDelegate handlerBinder = null;
+            for (var i = 0; i < _actionDescriptor.HandlerMethods.Count; i++)
```

Comment:
This is probably worth a comment, we end up doing two cache lookups, but it mostly seems fine and it would change a lot of stuff to reduce it to a single lookup